packages/@aws-cdk/cdk/lib/cloudformation/stack.ts

```diff
@@ -417,6 +418,38 @@ export class Stack extends Construct {
     }
   }
 
+  protected synthesize(session: ISynthesisSession): void {
+    const account = this.env.account || 'unknown-account';
+    const region = this.env.region || 'unknown-region';
+    const template = `${this.node.id}.template.json`;
+
+    // write the CloudFormation template as a JSON file
+    session.store.writeJson(template, this.toCloudFormation());
+
+    const artifact: cxapi.Artifact = {
```

Comment:
So it looks like staging will use the same format as cloud assembly? How would I communicate something like 'build this folder'? If it's an `Artifact`, then how do we plan on distinguishing between synth-only artifacts and cloud-assembly artifacts?

Reply:
The output of `App.run()` is an intermediate step towards a cloud assembly. I've added a `build: any` to `Artifact` as a placeholder, with the idea that the toolkit will use this information to build the intermediate artifact and produce the final assembly.

I am not in love with this design, but maybe it's the pragmatic thing to do, especially given the simple case (CloudFormation template) is a no-op between staging=>assembly.

Happy to hear your thoughts. Maybe we can start with this and iterate and see where it takes us.

Reply:
Thinking about this some more. I think you are right... Something is still not clicking. Crap.

NEED MORE THINKING.

Reply:
Okay, here's where I am at right now.

Let's say that `app.run()` produces two files: `manifest.json`, which is the cloud assembly manifest proper and `build.json` which build instructions for the toolkit.

Let's assume that `manifest.json` is not touched by the toolkit and that the build system (which is not modeled yet), will produce outputs that are pre-referenced by the emitted manifest.

As you can see in the example below, the artifact `MyHandlerCode` is of type `aws:s3:object` (because this is the resource that will eventually be created) and the `filePath` property points to `file://3cdf5efc05e1254da7d751257c344e74.zip`, which doesn't exist when `app.run()` ends. The file location is always relative to the assembly root.

```json
{
  "artifacts": {
    "MyStack": {
      "type": "aws:cloudformation:stack",
      "environment": "aws://account/region",
      "properties": {
        "name": "my-stack-name",
        "parameters": {
          "MyHandlerLocation": "s3://${MyHandlerCode.bucket}/${MyHandlerCode.key}"
        },
        "template": "file://${assembly}/my-stack.template.json"
      }
    },
    "MyHandlerCode": {
      "type": "aws:s3:object",
      "properties": {
        "filePath": "file://3cdf5efc05e1254da7d751257c344e74.zip"
      }
    }
  }
}
```

Completely decoupled from that, we also emit `build.json` (which is not modeled yet) but basically tells the toolkit how to produce `3cdf5efc05e1254da7d751257c344e74.zip`. This is where we can get wild with your declarative build system.

So, for the above example, let's say the handler is a Java project and we use lambda builders, then `build.json` will model the following process:

1. (future) Generate some magical code into the project directory (only referenced in `build.json`, not needed by `manifest.json`).
2. Run lambda-builders-java for the project directory inside the docker image
3. Zip up the resulting output (say it's not zipped by lambda-builders, I don't know)
4. Output to `3cdf5efc05e1254da7d751257c344e74.zip`

It is important that artifacts and build commands are completely decoupled, and only connect when a build output is used as an artifact's input. For example, the build output name is based on the content hash of the project directory, so if multiple assets use the same project directory, it's the same build output.